### ISSUE TITLE
Do not automatically apply SIP correction to alternative WCSes if SIP suffix is missing

### ISSUE THREAD

**Issue by Miladiouss** [astropy/astropy#13105]
### Description

If SIP coefficients are present in the header, Astropy applies it to all WCSes. If `CTYPE` of the WCS in hand doesn't have a `-SIP` suffix, Astropy will generate a warning and apply the SIP correction. It may be a valid approach to apply SIP to the primary WCS regardless of `-SIP` in `CTYPE`, however, Astropy should not automatically apply SIP correction to alternative WCSes if they don't have the `-SIP` ending in their `CTYPE` and it should not produce any warnings. This warning and default behavior has been a source of confiusion in the past (e.g. [issue 9239](https://github.com/astropy/astropy/issues/9239)).

### Steps to Reproduce

1. The following script will generate a minimalist FITS file to reproduce the described behavior in the next step:
 
```python
#! /usr/bin/env python
from astropy.io import fits
import numpy as np
from pathlib import Path
produtsPath = Path().parent / 'products'
produtsPath.mkdir(exist_ok=True)
def get_image_data(npix=99):
 y = np.arange(1, 1 + npix, dtype=np.float32)
 x = y / 100
 x_mesh, y_mesh = np.meshgrid(x, y)
 return x_mesh + y_mesh
def construct_sample_hdul(npix=99):
 data = get_image_data(npix=npix)
 hdul = fits.HDUList([fits.PrimaryHDU()])
 hdul.append(fits.ImageHDU(data=data, name='DATA'))
 hdr = hdul[1].header
 hdr['WCSAXES'] = 2
 hdr['CTYPE1 '] = 'RA---SIN-SIP'
 hdr['CTYPE2 '] = 'DEC--SIN-SIP'
 hdr['CUNIT1 '] = 'deg'
 hdr['CUNIT2 '] = 'deg'
 hdr['CRPIX1 '] = 1 # 1.5 + npix / 2
 hdr['CRPIX2 '] = 1 # 1.5 + npix / 2
 hdr['CRVAL1 '] = 0
 hdr['CRVAL2 '] = 90
 hdr['CDELT1 '] = round(6.2 / 3600, 5)
 hdr['CDELT2 '] = round(6.2 / 3600, 5)
 # i_j: output_input
 hdr['PC1_1 '] = 1
 hdr['PC1_2 '] = 0
 hdr['PC2_1 '] = 0
 hdr['PC2_2 '] = 1
 # ----------------
 # SIP coefficients
 a = 1e-5
 hdr['A_ORDER'] = 2
 hdr['A_0_2 '] = a
 hdr['A_1_1 '] = a
 hdr['A_2_0 '] = a
 hdr['B_ORDER'] = 2
 hdr['B_0_2 '] = a
 hdr['B_1_1 '] = a
 hdr['B_2_0 '] = a
 # Inverse SIP coefficients
 hdr['AP_ORDER'] = 2
 hdr['AP_0_2 '] = a
 hdr['AP_1_1 '] = a
 hdr['AP_2_0 '] = a
 hdr['BP_ORDER'] = 2
 hdr['BP_0_2 '] = a
 hdr['BP_1_1 '] = a
 hdr['BP_2_0 '] = a
 # ----------------
 hdr['WCSAXESW'] = 2
 hdr['CTYPE1W '] = 'WAVE'
 hdr['CTYPE2W '] = 'WAVE'
 hdr['CUNIT1W '] = 'm'
 hdr['CUNIT2W '] = 'm'
 hdr['CRPIX1W '] = 0
 hdr['CRPIX2W '] = 0
 hdr['CRVAL1W '] = 0
 hdr['CRVAL2W '] = 0
 hdr['CDELT1W '] = 1
 hdr['CDELT2W '] = 1
 # i_j: output_input
 hdr['PC1_1W '] = 1
 hdr['PC1_2W '] = 0
 hdr['PC2_1W '] = 0
 hdr['PC2_2W '] = 1
 return hdul
if __name__ == '__main__':
 hdul = construct_sample_hdul()
 hdul.writeto(produtsPath / 'minimal.fits', overwrite=True)
```
2. Initialize and read file header:

```python
# Standar Dependencies
import numpy as np
from astropy.io import fits
from astropy.wcs import WCS
import warnings
from pathlib import Path

# Notebook
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

# Read File
detector_id = 2
produtsPath = Path().parent / 'products'
# filePath = produtsPath / f'realistic_lvf_wcs_2p2d_array-{detector_id}.fits'
filePath = produtsPath / f'minimal.fits'
hdul = fits.open(filePath)

hdr = hdul[1].header
```

3. Reconstruct the alternative WCS _without_ suppressing SIP. To suppress, uncomment `wcs.sip = None`. 

```python

# Create a wcs object
with warnings.catch_warnings():
    warnings.filterwarnings(
        'ignore', message='.*The WCS transformation has more axes.*')
    wcs = WCS(hdr, hdul, 'W')
    # !!! toggle next line
    # wcs.sip = None
```
4. Depending on whether SIP was manually suppressed or not, the following pixel to world transformations would be different.

```python
# Query wavelength and bandwidth for a set of pixel coordinates
if True:
    b = 99
    x1 = [0, b, b, 0]
    y1 = [0, 0, b, b]
    corner_wl_out, corner_bandwidth_out = wcs.pixel_to_world(x1, y1)

    # Generate a report
    print(f'\nFile was read: {filePath.stem}')
    print('Input x, y Pixel Coord (counting from 0, 0):')
    print(x1, y1)
    print('Output Wavelength:')
    print(corner_wl_out)
    print('Output Bandwidth:')
    print(corner_bandwidth_out)
    print('\n------------------------------------\n')
```
Output with SIP suppression:

```text
File was read: minimal
Input x, y Pixel Coord (counting from 0, 0):
[0, 99, 99, 0] [0, 0, 99, 99]
Output Wavelength:
[  1. 100. 100.   1.] m
Output Bandwidth:
[  1.   1. 100. 100.] m
```

Output without SIP suppression:

```text
File was read: minimal
Input x, y Pixel Coord (counting from 0, 0):
[0, 99, 99, 0] [0, 0, 99, 99]
Output Wavelength:
[  1.00003 100.10101 100.3       1.10101] m
Output Bandwidth:
[  1.00003   1.10101 100.3     100.10101] m
```

This warning is generated regardless:

```text
INFO: 
                Inconsistent SIP distortion information is present in the FITS header and the WCS object:
                SIP coefficients were detected, but CTYPE is missing a "-SIP" suffix.
                astropy.wcs is using the SIP distortion coefficients,
                therefore the coordinates calculated here might be incorrect.

                If you do not want to apply the SIP distortion coefficients,
                please remove the SIP coefficients from the FITS header or the
                WCS object.  As an example, if the image is already distortion-corrected
                (e.g., drizzled) then distortion components should not apply and the SIP
                coefficients should be removed.

                While the SIP distortion coefficients are being applied here, if that was indeed the intent,
                for consistency please append "-SIP" to the CTYPE in the FITS header or the WCS object.

                 [astropy.wcs.wcs]
```

### Desired Behavior

No warning, no SIP correction as default behavior for alternative WCSes if `CTYPE` is missing a `-SIP` suffix. 

### System Details

```text
Linux-5.16.15-76051615-generic-x86_64-with-glibc2.34
Python 3.9.7 | packaged by conda-forge | (default, Sep  2 2021, 17:58:34) 
[GCC 9.4.0]
Numpy 1.22.3
pyerfa 2.0.0.1
astropy 5.0.4
Scipy 1.8.0
Matplotlib 3.5.1
```

**Comment by astrofrog** [comment]
I agree with this, and I am going to open a PR with a suggested fix.

**Comment by astrofrog** [comment]
Actually before rushing into this - relevant and long discussion here: https://github.com/astropy/astropy/issues/4803

**Comment by astrofrog** [comment]
@nden @mcara - I'm wondering whether it might be time to revisit this? Has anything changed in terms of WCS/SIP standards in the last 7 years since our discussion  in #4803? Is there a 'right' answer for how we should deal with these cases? If we want to switch to use -SIP in CTYPE as a toggle, we could introduce a warning in the upcoming 5.1 to say that the default behavior will change in future?

**Comment by nden** [comment]
Is the above example realistic? I'll assume it's not or otherwise I don't understand it. And if realistic then there are other problems - SIP does not work with spectral WCS.

Ignoring the above, I think an image is either distorted or undistorted. And because of this a distortion model applies to all alternate WCSs. This behavior is consistent with wcslib. Changing it will mean getting different results when running wcslib and astropy.

The problem is these are conventions that appeared and have been adopted at different times and their relation is not fully defined. Obviously, because of the 8 char FITS keyword limitation it's not possible to have multiple distortions in a header. What is proposed in this issue is also not fully defined, why apply to the primary but not to the alternate?

So I guess I disagree at a fundamental level but perhaps I don't understand the use case.
Of course it is possible to ignore not just the SIP but all distortions by running WCS.wcs_pix2world(), if desired.

**Comment by Miladiouss** [comment]
> Is the above example realistic? I'll assume it's not or otherwise I don't understand it. And if realistic then there are other problems - SIP does not work with spectral WCS.
> 
> Ignoring the above, I think an image is either distorted or undistorted. And because of this a distortion model applies to all alternate WCSs. This behavior is consistent with wcslib. Changing it will mean getting different results when running wcslib and astropy.

@nden - the NASA SPHEREx mission, for instance, uses a linear variable filter where each pixel has a different bandpass. We like to use an alternative WCS (`CTYPE`: 2-D `WAVE-TAB`) to represent wavelength and bandwidth for each pixel and they should not be distorted since pixel coordinates do not transform to sky coordinates in this case. 

> The problem is these are conventions that appeared and have been adopted at different times and their relation is not fully defined. Obviously, because of the 8 char FITS keyword limitation it's not possible to have multiple distortions in a header. What is proposed in this issue is also not fully defined, why apply to the primary but not to the alternate?

The reason I suggested the default astropy behavior should remain the same for the primary WCS is to avoid causing too much chaos for other users. Otherwise, I agree with @astrofrog that the definition is clear and consistency among all WCes is very appealing.
And yes, it is sad that there could be only one set of SIP distortion in the header with the current standards. 

> So I guess I disagree at a fundamental level but perhaps I don't understand the use case. Of course it is possible to ignore not just the SIP but all distortions by running WCS.wcs_pix2world(), if desired.

This is our current solution, however, we think general SPHEREx users would miss suppressing SIP distortions which in this case the discrepancies are non-negligible.

**Comment by mcara** [comment]
I have checked the behavior of the WCSLIB and it is consistent with the implementation in the `astropy.wcs`: SIP is loaded if SIP coefficients are found in the header regardless of whether `CTYPE` ends in `-SIP`. A long time ago [I also thought that "the definition was clear"](https://github.com/astropy/astropy/issues/5145#issuecomment-238456264) but [I no longer think so](https://github.com/astropy/astropy/issues/4803#issuecomment-214363682). Maybe that is how they [SIP authors] would have liked it but since appending `-SIP` to `CTYPE` was not supported by the "standard" it was mostly ignored (including by DS9 - that was my understanding - see https://github.com/astropy/astropy/issues/5145#issuecomment-238393530). In any case, this topic has so many different opinions and nothing is really clear about it. Implementing your proposed behavior (treat alt-wcs differently from the primary) would only increase confusion and very likely break existing code. Therefore I disagree with proposed change.

There are a couple of other points that I would like to make...

> The NASA SPHEREx mission, for instance, uses a linear variable filter where each pixel has a different bandpass. We like to use an alternative WCS (CTYPE: 2-D WAVE-TAB) to represent wavelength and bandwidth for each pixel and they should not be distorted since pixel coordinates do not transform to sky coordinates in this case.

It may be too late to do anything about it, but using 'WAVE-TAB' to store each pixel's bandwidth (as an example) seems an extreme and unnatural use of the WCS. This particular usage I would have implemented as another FITS extension in the HDU List and simply have another data array that would store the bandwidth. I think this would have been clear and without need to implement tricks to deal with WCS.

> ... we think general SPHEREx users would miss suppressing SIP distortions which in this case the discrepancies are non-negligible.

Since the use of 'WAVE-TAB' to encode non-wcs-related pixel characteristics is such a unique twist, I think this feature should be thoroughly documented and users be warned on what to do with such a WCS (i.e., strip SIP from it).

-------------------------

However, I do have another concern: '-SIP' is applicable only to celestial coordinates and therefore, *I wonder whether it is a bug to load SIP at all when there are no celestial axes in a WCS* (like the alternative WCS 'W' in the example above)?

**Comment by nden** [comment]
@mcara I agree, we can put in a check if the WCS is of type imaging when loading the SIP distortion and not load it if not imaging. It's another unconventional use of FITS and WCS, perhaps it's time for a new format ;-)

**Comment by Miladiouss** [comment]
@mcara - I understand your concern and perhaps all WCSes should behave the same way. I think @mcara's proposal is good one :+1: 

@mcara - this is not related to this issue but I'll mention it for completeness:
We do want to encode bandpass data with WCS, and WAV-TAB turn out to be the most efficient way to encode linear variable filters. We only use a 6x6 grid and WCS interpolates between the grid points with errors < .5% (this is already working with astropy). We are adding the same features to FireFly and trying to get DS9 onboard. Hopefully, other future projects would use WCS to compactly express pixel information if the right infrastructure is there.

**Comment by gpdf** [comment]
Just to get it in the record - the registered convention definition for SIP is here:
http://fits.gsfc.nasa.gov/registry/sip/SIP_distortion_v1_0.pdf

There are additional clarifications in this paper that were requested by the FITS review process.  I don't think any of the changes address issues relevant to this thread, though.  

It does seem clear that the authors intended "-SIP" to be the operational flag, and this did pass the FITS review process for a registered convention.

**Comment by gpdf** [comment]
There are some use cases in the Rubin Observatory system where it would not be appropriate for distortions to apply to alternate WCSes - we are able to use alternate WCSes to provide alternate representations of the physical coordinates of the detectors in the camera mosaic - supporting numbering systems at different levels of the assembly hierarchy.

Optical distortions of the telescope are irrelevant to these.

**Comment by mcara** [comment]
About 7 years ago there were numerous discussions about the SIP issue. Initially I thought that original intention by SIP authors was to have `'-SIP'` suffix in `CTYPE` to be used as a "trigger" for turning SIP distortion model ON/OFF. However, since by 2015 (time of my comments) both `astropy` and `wcslib` loaded SIP coefficients regardless of the `CTYPE` value,  and see citation in https://github.com/astropy/astropy/issues/4387#issuecomment-163675135 - this did not help me convince people to support using `CTYPE` as a switch for SIP.

In fact, this issue is very similar to https://github.com/astropy/astropy/issues/5801 at least at the level of usage but the decision was already made to load SIP model regardless of the `CTYPE` value.

> It does seem clear that the authors intended "-SIP" to be the operational flag, and this did pass the FITS review process for a registered convention.

With  regard to this, it may be helpful to invoke @stargaser. Maybe he could provide his perspective both from original intention and possibly to the current situation (consistency and backward compatibility point of view?)

IMO, it does not matter what they intended if extra long `CTYPE` (like `'RA---TAN-SIP'`) was not supported at the time when the initial paper came out (is that true @stargaser?). DS9 could not read it and then `astropy` and WCSLIB probably decided to load SIP coefficients anyway (regardless of the presence of `-SIP` in `CTYPE`) - that is just my understanding which could be wrong since I did not work on these packages at the time SIP papers came out.

I will copy/paste here a part of the text from my above linked comment:

> From a recent paper by David L. Shupe et al. ["More Flexibility in Representing Geometric Distortion in Astronomical Images" (2012)](http://authors.library.caltech.edu/36801/1/84511M.pdf): _"For SIP, the “trigger” is the addition of “-SIP” to the existing CTYPE1 and CTYPE2 values. This was proposed during the now-defunct effort to define “Paper IV” on distortion as part of the FITS standard. **Regrettably, this fine idea was later discarded and the FITS standard now states that CTYPEn should be eight characters only.**_

All this discussion is simply to show how complicated things are with regard to these extra 4 characters (`'-SIP'`) in the `CTYPE`. Now we have what we have and the question is how to fix it in a reasonable way.

**Comment by mcara** [comment]
Going forward, I think we need to make a couple of decisions such as:

1. Would there be any backward compatibility issues (i.e., did any observatory create FITS files without `-SIP` in `CTYPE` but they expected WCS libraries to load SIP coefficients from the header)?
2. How do we implement support for using `CTYPE` as a switch for SIP? How do we deal with these compatibility issues?

I do not know how to find out first point. Should we ask this via a poll or something similar? With regard to the second point:

- If we have a very good indication that no observatory ever used SIP in this way (i.e., not have `-SIP` in `CTYPE` but expected SIP to be loaded from SIP coefficients), then I think we should simply change the way we handle SIP and just start using `CTYPE` as a switch.
- Maybe, to extra cautious, we should introduce a deprecation period for the old usage?
- Or, do we support a software switch to control the `CTYPE` interpretation, like in https://github.com/astropy/astropy/pull/13155? However, I still do not like having a switch in the constructor - see my comments in https://github.com/astropy/astropy/pull/13155#pullrequestreview-950270226 and https://github.com/astropy/astropy/pull/13155#issuecomment-1106670682. **IF** we are going to support both (old and "new" treatment of `-SIP` suffix in `CTYPE`) then I would prefer to do that via a configuration file. I think we should try to preserve the "archival" spirit of the FITS format and not allow ambiguous interpretations of the content. The fact that this is in a configuration setting rather than in a WCS constructor would suggest more of a backward compatibility issue rather than supporting various interpretations of `CTYPE`.

I think the decision made should be agreed upon by a significant majority and once it is made, we must promise to never revisit this topic. 😄 

Aster several conversations that I had, it appears all HST images that intended to use SIP distortions should have had `'-SIP'` in `CTYPE` so there should not be any backward compatibility issues for STScI. I hope @nden can back me up.

It would be great if we could get more feedback on this issue from other observatories.

**Comment by mcara** [comment]
@gpdf With regard to https://github.com/astropy/astropy/issues/13105#issuecomment-1110171999, while I understand second point and the convenience of displaying values in DS9, it can be dealt with via loading each extension into individual frames in DS9 or opening files as cubes. Even then, maybe these wavelength axes could be the 3rd and 4th axes in the primary WCS instead of ALT WCS and then having users select the WCS they want to see? just like in Table 11 in [Paper III](http://www.atnf.csiro.au/people/mcalabre/WCS/scs.pdf)? [and maybe it is to late to make any changes now] 

With regard to the first point,  _"The wavelength model is inherently two-dimensional"_ maybe I am missing something: the section 6.1.1 of  the [Paper III](http://www.atnf.csiro.au/people/mcalabre/WCS/scs.pdf), starts with _"Consider first the case of a single (one-dimensional) coordinate axis."_ Our unit tests has FITS file with one -TAB axis and I think WCSLIB too has unit tests with a single -TAB axis.
